lib/rpmte.cc

```diff
@@ -151,17 +151,37 @@ static int addTE(rpmte p, Header h, fnpyKey key, rpmRelocation * relocs)
     p->release = headerGetAsString(h, RPMTAG_RELEASE);
 
     /* name, version and release are required in all packages */
-    if (p->name == NULL || p->version == NULL || p->release == NULL)
-	goto exit;
+    for (auto [value, name] : std::vector<std::pair<char *, std::string>>
+	     {{p->name, "name"},
+	      {p->version, "version"},
+	      {p->release, "release"}}) {
+	if (value == NULL) {
+	    rpmlog(RPMLOG_ERR, _("Package is missing %s\n"), name.c_str());
+	    goto exit;
+	}
+    }
 
     p->epoch = headerGetAsString(h, RPMTAG_EPOCH);
 
     p->arch = headerGetAsString(h, RPMTAG_ARCH);
     p->os = headerGetAsString(h, RPMTAG_OS);
 
-    /* gpg-pubkey's dont have os or arch (sigh), for others they are required */
-    if (!rstreq(p->name, "gpg-pubkey") && (p->arch == NULL || p->os == NULL))
+    if (p->arch == NULL || p->os == NULL) {
+	if (p->type == TR_REMOVED && rstreq(p->name, "gpg-pubkey")) {
+	    rpmlog(RPMLOG_WARNING, "Erasing gpg-pubkey packages is deprecated! Use rpmkeys.\n");
+	} else {
+	    if (p->arch == NULL)
+		rpmlog(RPMLOG_ERR, _("Package is missing %s\n"), "arch");
+	    if (p->os == NULL)
+		rpmlog(RPMLOG_ERR, _("Package is missing %s\n"), "os");
```

Comment:
If you want to add an error message, and one central "malformed package" style thing at the end of addTE() on error, that'll cover all the existing sanity checks too. There are a hundred thousand ways things can go wrong and we can't have specific messages for each.